services/rpcfilters/latest_block_changed_event_test.go

```diff
@@ -0,0 +1,163 @@
+package rpcfilters
+
+import (
+	"math/big"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/stretchr/testify/assert"
+)
+
+type latestBlockProviderTest struct {
+	BlockFunc func() (blockInfo, error)
+}
+
+func (p latestBlockProviderTest) GetLatestBlock() (blockInfo, error) {
+	return p.BlockFunc()
+}
+
+func TestEventSubscribe(t *testing.T) {
+	counter := 0
+
+	hashes := []common.Hash{common.HexToHash("0xAA"), common.HexToHash("0xBB"), common.HexToHash("0xCC")}
+
+	f := func() (blockInfo, error) {
+		counter++
+		number := big.NewInt(int64(counter))
+		if counter > len(hashes) {
+			counter = len(hashes)
+		}
+		return blockInfo{hashes[counter-1], hexutil.Bytes(number.Bytes())}, nil
+	}
+
+	testEventSubscribe(t, f, hashes)
+}
+
+func TestZeroSubsciptionsOptimization(t *testing.T) {
+	counter := 0
+	hash := common.HexToHash("0xFF")
+
+	f := func() (blockInfo, error) {
+		counter++
+		number := big.NewInt(1)
+		return blockInfo{hash, hexutil.Bytes(number.Bytes())}, nil
+	}
+
+	event := newLatestBlockChangedEvent(latestBlockProviderTest{f})
+	event.tickerPeriod = time.Millisecond
+
+	assert.NoError(t, event.Start())
+	defer func() { assert.NoError(t, event.Stop()) }()
+
+	// let the ticker to call ~10 times
+	time.Sleep(10 * time.Millisecond)
+
+	// check that our provider function wasn't called when there are no subscribers to it
+	assert.Equal(t, 0, counter)
+
+	// subscribing an event, checking that it works
+	id, channel := event.Subscribe()
+
+	timeout := time.After(1 * time.Second)
+	select {
+	case receivedHash := <-channel:
+		assert.Equal(t, hash, receivedHash)
+	case <-timeout:
+		assert.Fail(t, "timeout")
+	}
+
+	event.Unsubscribe(id)
+
+	// check that our function was called multiple times
+	assert.True(t, counter > 0)
+	counterValue := counter
+
+	// let the ticker to call ~10 times
+	time.Sleep(10 * time.Millisecond)
+
+	// check that our provider function wasn't called when there are no subscribers to it
+	assert.Equal(t, counterValue, counter)
+}
+
+func TestMultipleSubscribe(t *testing.T) {
+	hash := common.HexToHash("0xFF")
+
+	f := func() (blockInfo, error) {
+		number := big.NewInt(1)
+		return blockInfo{hash, hexutil.Bytes(number.Bytes())}, nil
+	}
+
+	event := newLatestBlockChangedEvent(latestBlockProviderTest{f})
+	event.tickerPeriod = time.Millisecond
+
+	wg := sync.WaitGroup{}
+	wg.Add(3)
+
+	testFunc := func() {
+		testEvent(t, event, []common.Hash{hash})
+		wg.Done()
+	}
+
+	go testFunc()
```

Comment:
for loop?

Reply:
that is a simple way to have 3 subscribers at the same time to make sure the all of them received a hash

Reply:
You can tie it up with a variable and a loop:
```go
repeat := 3
wg.Add(repeat)
for i := 0; i < repeat; i++ {
  go testFunc()
}
```